tools/update_runnablejobs.py

```diff
@@ -191,59 +260,88 @@ def add_jobs_to_jobpriority(new_data=None, priority=1, timeout=0, set_expired=Fa
-        platform = job["platform"]
-        testtype = job["testtype"]
-
-        found = False
-        found_id = None
+    # TODO: write test for this
+    # When the table is empty it means that we're starting the system for the first time
+    # and we're going to use different default values
+    map = {}
+    if not len(db_data) == 0:
+        priority, timeout = 1, 0
+        # Using %Y-%m-%d fixes this issue:
+        # Warning: Incorrect date value: '2016-10-28 17:36:58.153265' for column 'expires' at row 1
+        expiration_date = (datetime.datetime.now() + datetime.timedelta(days=14)).strftime("%Y-%m-%d")
+        # Creating this data structure which will reduce how many times we iterate through the DB rows
         for row in db_data:
-            if (row[1] == testtype and
-                    row[3] == platform and
-                    row[2] == job["platform_option"]):
-                # TODO: what if we have a race condition with two identical jobs
-                # verify the build system type is the same, or make it *
-                found = True
-                if row[7] != "*" and _buildsystem != row[7]:
+            key = tuple(row[1:4])
+            # This is guaranteed by a unique composite index for these 3 fields in models.py
+            assert key not in map,\
+                '"{}" should be a unique row and that is unexpected.'.format(key)
+            # (testtype, buildtype, platform)
+            map[key] = {'pk': row[0], 'build_system_type': row[7]}
+    else:
+        priority, timeout, expiration_date = 5, 5400, None
+
+    total_jobs, new_jobs, failed_changes, updated_jobs = len(data), 0, 0, 0
+    # Loop through sanitized jobs, add new jobs and update the build system if needed
+    for job in data:
+        _buildsystem = job["build_system_type"]
+        key = _unique_key(job)
+        if key in map:
+            # We already know about this job, we might need to update the build system
+            row_build_system_type = map[key]['build_system_type']
+
+            if row_build_system_type == '*' or _buildsystem == '*':
+                # We don't need to update anything
+                pass
+            else:
+                # We're seeing the job again but for another build system (e.g. buildbot vs
+                # taskcluster). We need to change it to '*'
+                if row_build_system_type != _buildsystem:
                     _buildsystem = "*"
-                    found_id = row[0]
 
-        # We have new jobs from runnablejobs to add to our master list
-        if not found:
-            _expired = None
-            if set_expired:
-                # set _expired = today + 14 days
-                # TODO: write test for it
-                _expired = "%s" % (datetime.datetime.now() + datetime.timedelta(days=14))
+                # Compare original build system to the new one
+                if _buildsystem != job['build_system_type']:
```

Comment:
you will only be in this if statement if you hit the condition in line 298, lets just have a single if statement here.